Feature/ms upgrade

fahado-aws opened the pull request:
*Issue #, if available:*

*Description of changes:*

Add new modes for upgrading to multiple streams
This change adds two new modes in which a KCL application can run -
SINGLE_STREAM_COMPATIBLE_MODE and SINGLE_STREAM_UPGRADE_MODE.
These will help upgrade an application that's using a single stream to
an application that can process multiple streams.

Add support for upgrading to multiple streams

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

adpeace commented on amazon-kinesis-client/src/main/java/software/amazon/kinesis/coordinator/PeriodicShardSyncManager.java:
```diff
@@ -277,9 +277,11 @@ private void runShardSync() {
             final Set<StreamIdentifier> streamIdentifiersToFilter)
             throws DependencyException, ProvisionedThroughputException, InvalidStateException {
-        final List<Lease> leases = leaseRefresher.listLeases();
+        final List<Lease> leases = leaseRefresher.listLeases();        
```
remove unintended whitespace change

adpeace commented on amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/LeaseManagementFactory.java:
```diff
@@ -26,6 +29,11 @@
 
+    default LeaseCoordinator createLeaseCoordinator(MetricsFactory metricsFactory, 
+        Map<StreamIdentifier, StreamConfig> streamConfigMap) {
+            throw new UnsupportedOperationException();
```
wrong whitespace

adpeace commented on amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/LeaseManagementFactory.java:
```diff
@@ -47,4 +55,9 @@ default ShardDetector createShardDetector(StreamConfig streamConfig) {
 
+    default LeaseCleanupManager createLeaseCleanupManager(MetricsFactory metricsFactory, 
+        Map<StreamIdentifier, StreamConfig> streamConfigMap) {
+            throw new UnsupportedOperationException();
```
wrong whitespace

adpeace commented on amazon-kinesis-client/src/main/java/software/amazon/kinesis/coordinator/Scheduler.java:
```diff
@@ -153,6 +153,7 @@ public class Scheduler implements Runnable {
     private final long failoverTimeMillis;
     private final long taskBackoffTimeMillis;
     private final boolean isMultiStreamMode;
+    private final StreamProcessingMode streamProcessingMode;
```
I'm wondering whether this should replace isMultiStreamMode instead of being extra, as we have two things tracking the state now and it's hard to tell if a usage of isMultiStreamMode should have been updated or is ok.

adpeace commented on amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/dynamodb/DynamoDBLeaseCoordinator.java:
```diff
@@ -194,14 +197,57 @@ public DynamoDBLeaseCoordinator(final LeaseRefresher leaseRefresher,
+            null, null);
+    }
+    /**
+     * Constructor.
```
Can we add a better description of what the constructor is for?

adpeace commented on amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/dynamodb/DynamoDBLeaseCoordinator.java:
```diff
@@ -194,14 +197,57 @@ public DynamoDBLeaseCoordinator(final LeaseRefresher leaseRefresher,
+     *            Initial dynamodb lease table write iops if creating the lease table
+     * @param metricsFactory
+     *            Used to publish metrics about lease operations
+     */
```
misses new params

adpeace commented on amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/dynamodb/DynamoDBLeaseRefresher.java:
```diff
@@ -707,6 +713,55 @@ public void deleteLease(@NonNull final Lease lease)
+        Put createNewLease = Put.builder()
+                .tableName(table)
+                .item(serializer.toDynamoRecord(newLease))
+                .conditionExpression(String.format("attribute_not_exists(%s)", "leaseKey"))
```
Should move out leaseKey into a referenced attribute name here:
```
Put createNewLease = Put.builder()
        .tableName(table)
        .item(serializer.toDynamoRecord(newLease))
        .conditionExpression("attribute_not_exists(#leaseKey)")
        .expressionAttributeNames(ImmutableMap.of("#leaseKey", "leaseKey"))
        .build();
```

adpeace commented on amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/dynamodb/DynamoDBLeaseRefresher.java:
```diff
@@ -707,6 +713,55 @@ public void deleteLease(@NonNull final Lease lease)
+        TransactWriteItemsRequest request = TransactWriteItemsRequest.builder()
+            .transactItems(actions)
+            .build();
+            
```
nit: need to do trailing whitespace cleanup across all your updates.

adpeace commented on amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/LeaseRefresher.java:
```diff
@@ -243,4 +243,18 @@ default void updateLeaseWithMetaInfo(Lease lease, UpdateField updateField)
+     * @throws InvalidStateException
+     * @throws ProvisionedThroughputException
+     */
+    default boolean replaceLease(Lease oldLease, Lease newLease)
```
I see this pattern of using `UnsupportedOperationException` is used elsewhere; I'm wondering what the effect of it being thrown is and if it makes sense for an implementation to omit this?

adpeace commented on amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/dynamodb/DynamoDBLeaseRenewer.java:
```diff
@@ -364,14 +406,29 @@ public void addLeasesToRenew(Collection<Lease> newLeases) {
-             */
-            authoritativeLease.concurrencyToken(UUID.randomUUID());
-            ownedLeases.put(authoritativeLease.leaseKey(), authoritativeLease);
+            log.info("addLeasesToRenew - Checking stream processing mode: {}", streamProcessingMode);
```
This log might be quite noisy and I'm not sure how useful it is.  Maybe move it into `if` block?

adpeace commented on amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/dynamodb/DynamoDBLeaseRenewer.java:
```diff
@@ -364,14 +406,29 @@ public void addLeasesToRenew(Collection<Lease> newLeases) {
+                */
+                authoritativeLease.concurrencyToken(UUID.randomUUID());
+                ownedLeases.put(authoritativeLease.leaseKey(), authoritativeLease);
+            }
```
Not sure if this is better, but wondering if  this be structured as first transitioning the lease to multi-stream if needed, and then continuing the existing behaviour. This avoids having to maintain two code paths after the initial lease transition. E.g. (untested!):

```
            Lease workingLease = lease;

            /* Attempt to convert lease to multi-stream if needed: */
            if (streamProcessingMode == StreamProcessingMode.SINGLE_STREAM_UPGRADE_MODE 
                && !(lease instanceof MultiStreamLease)) {
                MultiStreamLease multiStreamLease = convertToMultiStreamLease(lease);
                if (leaseRefresher.replaceLease(lease, multiStreamLease)) {
                    log.info("Worker {} replaced single stream lease {} with multi-stream lease {}", workerIdentifier, lease, multiStreamLease);
                    ownedLeases.remove(lease);
                    workingLease = multiStreamLease;
                } else {
                    // skip this lease:
                    log.warn("Worker {} ignoring single-stream lease: {} because it could not be " + 
                        "replaced with multi-stream lease: {}", workerIdentifier, lease, multiStreamLease);
                    continue;
                }
            }

            Lease authoritativeLease = workingLease.copy();
            /*
             * Assign a concurrency token when we add this to the set of currently owned leases. This ensures that
             * every time we acquire a lease, it gets a new concurrency token.
             */
            authoritativeLease.concurrencyToken(UUID.randomUUID());
            ownedLeases.put(authoritativeLease.leaseKey(), authoritativeLease);
```

adpeace commented on amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/dynamodb/DynamoDBLeaseRenewer.java:
An idea to consider: Should we have a separate class for doing the lease upgrade that is passed in as a dependency to this one, and knows the stream identifier?  It seems a bit weird that this class that didn't used to know the stream map now does just for the purpose of doing the upgrade.

Or alternatively, rather than passing the whole map, maybe just pass the stream info to add to single-stream leases?

adpeace commented on amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/HierarchicalShardSyncer.java:
```diff
@@ -78,6 +79,10 @@ public class HierarchicalShardSyncer {
 
     private final DeletedStreamListProvider deletedStreamListProvider;
 
+    private final StreamProcessingMode streamProcessingMode;
```
Not actually used.  I think we should probably keep it and switch out all uses of isMultiStream for `streamProcessingMode == ...` so the code is as explicit as possible.

adpeace commented on amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/HierarchicalShardSyncer.java:
```diff
@@ -762,8 +780,8 @@ public int compare(final Lease lease1, final Lease lease2) {
     @VisibleForTesting
-    static class MultiStreamArgs {
-        private final Boolean isMultiStreamMode;
+    static class MultiStreamLeaseArgs {
```
curious what the reason for renaming is?

adpeace commented on amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/HierarchicalShardSyncer.java:
```diff
@@ -92,16 +97,29 @@ public HierarchicalShardSyncer(final boolean isMultiStreamMode, final String str
     }
 
-    private static final BiFunction<Lease, MultiStreamArgs, String> SHARD_ID_FROM_LEASE_DEDUCER =
+    private static final BiFunction<Lease, MultiStreamLeaseArgs, String> SHARD_ID_FROM_LEASE_DEDUCER =
```
No longer needs to be a `BiFunction` as the second argument isn't used.

adpeace commented on amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/HierarchicalShardSyncer.java:
```diff
@@ -152,8 +170,9 @@ public synchronized boolean checkAndCreateLeaseForNewShards(@NonNull final Shard
-        final MultiStreamArgs multiStreamArgs = new MultiStreamArgs(isMultiStreamMode, shardDetector.streamIdentifier());
+                leaseRefresher.listLeasesForStream(shardDetector.streamIdentifier()) : leaseRefresher.listLeases();        
+        final MultiStreamLeaseArgs multiStreamArgs = new MultiStreamLeaseArgs(createMultiStreamLease, 
+            shardDetector.streamIdentifier());
```
This whole args thing ... it seems to be a way to tell the lease renewers how to create a lease.  Should we just pass a lease factory instead?

adpeace commented on amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/LeaseCoordinator.java:
```diff
@@ -49,8 +49,9 @@ public interface LeaseCoordinator {
+     * @throws ProvisionedThroughputException 
      */
-    void runLeaseTaker() throws DependencyException, InvalidStateException;
+    void runLeaseTaker() throws DependencyException, InvalidStateException, ProvisionedThroughputException;
```
Seems wrong to throw a DDB-specific exception in this interface that (in theory...) isn't DDB-specific.  I haven't quite followed through why this changed, but I'd suggest rethrowing something relevant in the implementation as needed.

adpeace reviewed (commented):
Put some initial comments on here.  I haven't yet run this or looked at the test code or thought through the distributed failure modes, but wanted to get something out.  Some of the comments are ideas as I'm not an expert on this code base, please handle them as you see fit.

jahyunk committed:
Add supports of the new modes for integration test

jahyunk committed:
Add supports of running multiple consumers for integration test

jahyunk committed:
Add integration tests updating mode through recommended upgrade path

jahyunk committed:
Fix package definition for ReleaseCanaryMultiStreamPollingH2TestConfig

fahado-aws committed:
Add new modes for upgrading to multiple streams

This change adds two new modes in which a KCL application can run -
SINGLE_STREAM_COMPATIBLE_MODE and SINGLE_STREAM_UPGRADE_MODE.
These will help upgrade an application that's using a single stream to
an application that can process multiple streams.

fahado-aws committed:
Add support for upgrading to multiple streams

This change adds support to do a live upgrade from the application
processing a single stream to being able to process multiple streams.
This can be done through phased deployments by switching the
application through different stream processing modes.

This change also adds new metrics to assist in determining readiness
to move from one mode to another. The metrics are also there to detect
unexpected outcomes of the upgrade and operational issues.

jahyunk committed:
Add integration tests of entire scenarios updating mode

fahado-aws committed:
Remove shard syncer stream compability checks

fahado-aws committed:
Add metrics to count leases per type

fahado-aws committed:
Add metric to record the worker mode

fahado-aws committed:
Deprecate no-arg constructor

fahado-aws committed:
Add more tests

fahado-aws committed:
Add LeaseTaker tests

fahado-aws committed:
Add test

fahado-aws committed:
Add counter check on replace and operation metrics

fahado-aws committed:
Emit all mode metrics to facilitate averaging

fahado-aws committed:
Fix lease cleanup in SingleStreamCompatibleMode

This change updates lease cleanup logic to look at both types of leases
when attempting to find child shards for a parent lease that has reached
end of processing. A method is added to LeaseRefresher to retrieve a
lease using shard Id.

fahado-aws committed:
Fix lease creation during shard merge in new modes

This change updates consumer shutdown task to look at both types of
parent leases when attempting to create a child lease during a shard
merge. This is relevant for the new modes added to support live
migration to multi-stream processing.

fahado-aws committed:
Look up all types of leases when locating parents

fahado-aws committed:
Emit KCL version as a metric

fahado-aws committed:
Audit upgraded lease and generate metrics

This change adds a follow-up audit operation to ensure that the
upgraded lease can be read back correctly.

fahado-aws committed:
Add shardId dimension in lease upgrade operations

This change adds shardId as a dimension for both UpgradeLease
and AuditUpgradedLease operations. This is to assist with finding
leases that fail to get upgraded.
